detectron2/structures/boxes.py

```diff
@@ -144,12 +144,14 @@ def __init__(self, tensor: torch.Tensor):
         Args:
             tensor (Tensor[float]): a Nx4 matrix.  Each row is (x1, y1, x2, y2).
         """
-        device = tensor.device if isinstance(tensor, torch.Tensor) else torch.device("cpu")
-        tensor = torch.as_tensor(tensor, dtype=torch.float32, device=device)
+        if not isinstance(tensor, torch.Tensor):
+            tensor = torch.as_tensor(tensor, dtype=torch.float32, device=torch.device("cpu"))
+        else:
+            tensor = tensor.to(torch.float32)
         if tensor.numel() == 0:
             # Use reshape, so we don't end up creating a new tensor that does not depend on
             # the inputs (and consequently confuses jit)
-            tensor = tensor.reshape((-1, 4)).to(dtype=torch.float32, device=device)
+            tensor = tensor.reshape((-1, 4)).to(dtype=torch.float32)
```

Comment:
I think this change does not help: `to(dtype=float32)` will still hardcode a device in the `to` operation if I'm not mistaken.

Reply:
Originally I was thinking `.to(dtype=float32)` will hardcode a device, so that's why changed `boxes = boxes.to(deltas.dtype)`, but somehow it doesn't seem necessary. I reverted all the changed in `box_regression.py` and it still worked.

One thing I'm sure is that `tensor_a.to(tensor_b)` will cause hardcoded device, although it's an all-tensor interface.